File: database/import_data.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Script para importar datos CSV a MySQL
"""
import csv
import mysql.connector
import os
import sys

# Configuración de MySQL
DB_CONFIG = {
    'host': os.getenv('MYSQL_HOST', 'localhost'),
    'port': int(os.getenv('MYSQL_PORT', '3306')),
    'user': os.getenv('MYSQL_USER', 'mtto_user'),
    'password': os.getenv('MYSQL_PASSWORD', 'mtto_password'),
    'database': os.getenv('MYSQL_DATABASE', 'mtto_db'),
    'charset': 'utf8mb4',
    'autocommit': False
}

def connect_db():
    """Conectar a la base de datos"""
    try:
        conn = mysql.connector.connect(**DB_CONFIG)
        return conn
    except mysql.connector.Error as e:
        print(f"❌ Error conectando a MySQL: {e}")
        sys.exit(1)

def import_csv(conn, csv_file, table_name, columns, null_columns=None):
    """Importar un archivo CSV a una tabla"""
    null_columns = null_columns or []
    
    try:
        cursor = conn.cursor()
        
        # Leer CSV
        with open(csv_file, 'r', encoding='utf-8') as f:
            reader = csv.DictReader(f)
            rows = list(reader)
        
        if not rows:
            print(f"⚠️  {table_name}: Archivo vacío")
            return
        
        # Preparar INSERT
        placeholders = ', '.join(['%s'] * len(columns))
        columns_str = ', '.join(columns)
        query = f"INSERT INTO {table_name} ({columns_str}) VALUES ({placeholders})"
        
        # Insertar datos
        inserted = 0
        for row in rows:
            values = []
            for col in columns:
                val = row.get(col, '')
                # Manejar valores NULL
                if col in null_columns and (val == 'NULL' or val == ''):
                    values.append(None)
                elif val == '':
                    values.append(None)
                else:
                    values.append(val)
            
            try:
                cursor.execute(query, values)
                inserted += 1
            except mysql.connector.Error as e:
                print(f"⚠️  Error insertando fila en {table_name}: {e}")
                print(f"   Valores: {values}")
        
        conn.commit()
        print(f"✅ {table_name}: {inserted} registros insertados")
        cursor.close()
        
    except Exception as e:
        print(f"❌ Error importando {table_name}: {e}")
        conn.rollback()

def main():
    print("📊 Importando datos CSV a MySQL...")
    print("")
    
    conn = connect_db()
    cursor = conn.cursor()
    
    # Deshabilitar verificaciones de claves foráneas
    cursor.execute("SET FOREIGN_KEY_CHECKS = 0")
    
    data_dir = "database/data"
    
    # 1. Maquinaria
    import_csv(conn, f"{data_dir}/01_maquinaria.csv", "maquinaria", 
               ['id', 'codigo', 'nombre', 'marca', 'modelo', 'anio', 'estado', 'costo_adquisicion', 'horas_totales'])
    
    # 2. Personal
    import_csv(conn, f"{data_dir}/02_personal.csv", "personal",
               ['id', 'codigo', 'nombre_completo', 'ci', 'cargo', 'telefono', 'celular', 'tarifa_hora', 'estado'])
    
    # 3. Herramientas
    import_csv(conn, f"{data_dir}/03_herramientas.csv", "herramientas",
               ['id', 'codigo', 'nombre', 'marca', 'estado', 'categoria', 'costo'])
    
    # 4. Insumos
    import_csv(conn, f"{data_dir}/04_insumos.csv", "insumos",
               ['id', 'codigo', 'nombre', 'unidad', 'precio_unitario', 'cantidad', 'stock_minimo', 'categoria'])
    
    # 5. Planes de Mantenimiento (con NULLs)
    import_csv(conn, f"{data_dir}/05_planes_mantenimiento.csv", "planes_mantenimiento",
               ['id', 'maquinaria_id', 'nombre_plan', 'tipo_mantenimiento', 'tipo_plan', 'horas_operacion', 'intervalo_dias', 'descripcion', 'activo'],
               null_columns=['horas_operacion', 'intervalo_dias'])
    
    # 6. Actividades
    import_csv(conn, f"{data_dir}/06_actividades_mantenimiento.csv", "actividades_mantenimiento",
               ['id', 'plan_id', 'numero_orden', 'descripcion_componente', 'actividad', 'tiempo_min', 'tiempo_promedio', 'tiempo_max', 'costo_estimado'])
    
    # 7. Actividad Insumos (con NULLs)
    import_csv(conn, f"{data_dir}/07_actividad_insumos.csv", "actividad_insumos",
               ['id', 'actividad_id', 'insumo_id', 'cantidad', 'unidad', 'especificaciones', 'costo_unitario'],
               null_columns=['insumo_id'])
    
    # 8. Actividad Herramientas (con NULLs)
    import_csv(conn, f"{data_dir}/08_actividad_herramientas.csv", "actividad_herramientas",
               ['id', 'actividad_id', 'herramienta_id', 'cantidad', 'especificaciones'],
               null_columns=['herramienta_id'])
    
    # 9. Mantenimientos (con NULLs)
    import_csv(conn, f"{data_dir}/09_mantenimientos.csv", "mantenimientos",
               ['id', 'maquinaria_id', 'plan_id', 'tipo_mantenimiento', 'fecha_programada', 'fecha_ejecucion', 'horas_maquina', 'estado', 'observaciones', 'costo_mano_obra', 'costo_insumos'],
               null_columns=['plan_id', 'fecha_programada', 'fecha_ejecucion', 'observaciones'])
    
    # 10. Mantenimiento Personal
    import_csv(conn, f"{data_dir}/10_mantenimiento_personal.csv", "mantenimiento_personal",
               ['id', 'mantenimiento_id', 'personal_id', 'horas_trabajadas', 'tarifa_aplicada'])
    
    # 11. Mantenimiento Insumos
    import_csv(conn, f"{data_dir}/11_mantenimiento_insumos.csv", "mantenimiento_insumos",
               ['id', 'mantenimiento_id', 'insumo_id', 'cantidad_usada', 'unidad', 'precio_unitario'])
    
    # 12. Mantenimiento Actividades (con NULLs)
    import_csv(conn, f"{data_dir}/12_mantenimiento_actividades.csv", "mantenimiento_actividades",
               ['id', 'mantenimiento_id', 'actividad_id', 'descripcion', 'tiempo_real', 'completada', 'observaciones'],
               null_columns=['actividad_id', 'observaciones'])
    
    # Habilitar verificaciones de claves foráneas
    cursor.execute("SET FOREIGN_KEY_CHECKS = 1")
    
    # Verificar datos
    print("")
    print("📊 Verificando datos cargados...")
    cursor.execute("""
        SELECT 'Maquinaria' as tabla, COUNT(*) as registros FROM maquinaria
        UNION ALL SELECT 'Personal', COUNT(*) FROM personal
        UNION ALL SELECT 'Herramientas', COUNT(*) FROM herramientas
        UNION ALL SELECT 'Insumos', COUNT(*) FROM insumos
        UNION ALL SELECT 'Planes', COUNT(*) FROM planes_mantenimiento
        UNION ALL SELECT 'Actividades', COUNT(*) FROM actividades_mantenimiento
        UNION ALL SELECT 'Mantenimientos', COUNT(*) FROM mantenimientos
    """)
    
    results = cursor.fetchall()
    for row in results:
        print(f"   {row[0]}: {row[1]} registros")
    
    cursor.close()
    conn.close()
    
    print("")
    print("✅ Importación completada!")

if __name__ == "__main__":
    main()

File: database/import_data.js
// Script para importar datos CSV a MySQL
import fs from 'fs';
import mysql from 'mysql2/promise';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

// Configuración de MySQL
const dbConfig = {
    host: process.env.MYSQL_HOST || 'mysql',
    port: parseInt(process.env.MYSQL_PORT || '3306'),
    user: process.env.MYSQL_USER || 'mtto_user',
    password: process.env.MYSQL_PASSWORD || 'mtto_password',
    database: process.env.MYSQL_DATABASE || 'mtto_db',
    charset: 'utf8mb4',
    multipleStatements: true
};

function parseCSV(filePath) {
    const content = fs.readFileSync(filePath, 'utf-8');
    const lines = content.split('\n').filter(line => line.trim());
    const headers = lines[0].split(',').map(h => h.trim().replace(/^"|"$/g, ''));
    const rows = [];
    
    for (let i = 1; i < lines.length; i++) {
        const values = [];
        let current = '';
        let inQuotes = false;
        
        for (let j = 0; j < lines[i].length; j++) {
            const char = lines[i][j];
            if (char === '"') {
                inQuotes = !inQuotes;
            } else if (char === ',' && !inQuotes) {
                values.push(current.trim());
                current = '';
            } else {
                current += char;
            }
        }
        values.push(current.trim());
        
        const row = {};
        headers.forEach((header, idx) => {
            let val = values[idx] || '';
            val = val.replace(/^"|"$/g, '');
            // Manejar valores booleanos
            if (val === 'TRUE' || val === 'true' || val === '1') {
                row[header] = 1;
            } else if (val === 'FALSE' || val === 'false' || val === '0') {
                row[header] = 0;
            } else if (val === 'NULL' || val === '') {
                row[header] = null;
            } else {
                row[header] = val;
            }
        });
        rows.push(row);
    }
    
    return { headers, rows };
}

async function importCSV(conn, csvFile, tableName, nullColumns = []) {
    try {
        const { headers, rows } = parseCSV(csvFile);
        
        if (rows.length === 0) {
            console.log(`⚠️  ${tableName}: Archivo vacío`);
            return;
        }
        
        const placeholders = headers.map(() => '?').join(', ');
        const columns = headers.join(', ');
        const query = `INSERT INTO ${tableName} (${columns}) VALUES (${placeholders})`;
        
        let inserted = 0;
        for (const row of rows) {
            const values = headers.map(h => {
                const val = row[h];
                if (nullColumns.includes(h) && (val === null || val === 'NULL' || val === '')) {
                    return null;
                }
                return val;
            });
            
            try {
                // Validar que los costos no sean 0 para tablas de costos
                if (tableName.includes('mantenimiento') || tableName.includes('actividad') || tableName === 'maquinaria' || tableName === 'herramientas' || tableName === 'insumos') {
                    const costFields = ['costo_estimado', 'costo_mano_obra', 'costo_insumos', 'costo_total', 'costo_adquisicion', 'costo', 'precio_unitario'];
                    const hasZeroCost = headers.some((h, idx) => {
                        if (costFields.includes(h)) {
                            const val = values[idx];
                            return val === 0 || val === '0' || val === '0.00' || parseFloat(val) === 0;
                        }
                        return false;
                    });
                    if (hasZeroCost) {
                        console.log(`⚠️  Omitiendo fila en ${tableName}: contiene costo en cero`);
                        continue;
                    }
                }
                await conn.execute(query, values);
                inserted++;
            } catch (error) {
                console.log(`⚠️  Error insertando fila en ${tableName}: ${error.message}`);
            }
        }
        
        console.log(`✅ ${tableName}: ${inserted} registros insertados`);
    } catch (error) {
        console.log(`❌ Error importando ${tableName}: ${error.message}`);
    }
}

async function main() {
    console.log('📊 Importando datos CSV a MySQL...');
    console.log('');
    
    let conn;
    try {
        conn = await mysql.createConnection(dbConfig);
        await conn.execute('SET FOREIGN_KEY_CHECKS = 0');
        
        const dataDir = join(__dirname, 'data');
        
        // 1. Maquinaria
        await importCSV(conn, join(dataDir, '01_maquinaria.csv'), 'maquinaria');
        
        // 2. Personal
        await importCSV(conn, join(dataDir, '02_personal.csv'), 'personal');
        
        // 3. Herramientas
        await importCSV(conn, join(dataDir, '03_herramientas.csv'), 'herramientas');
        
        // 4. Insumos
        await importCSV(conn, join(dataDir, '04_insumos.csv'), 'insumos');
        
        // 5. Planes de Mantenimiento
        await importCSV(conn, join(dataDir, '05_planes_mantenimiento.csv'), 'planes_mantenimiento', ['horas_operacion', 'intervalo_dias']);
        
        // 6. Actividades
        await importCSV(conn, join(dataDir, '06_actividades_mantenimiento.csv'), 'actividades_mantenimiento');
        
        // 7. Actividad Insumos
        await importCSV(conn, join(dataDir, '07_actividad_insumos.csv'), 'actividad_insumos', ['insumo_id']);
        
        // 8. Actividad Herramientas
        await importCSV(conn, join(dataDir, '08_actividad_herramientas.csv'), 'actividad_herramientas', ['herramienta_id']);
        
        // 9. Mantenimientos
        await importCSV(conn, join(dataDir, '09_mantenimientos.csv'), 'mantenimientos', ['plan_id', 'fecha_programada', 'fecha_ejecucion', 'observaciones']);
        
        // 10. Mantenimiento Personal
        await importCSV(conn, join(dataDir, '10_mantenimiento_personal.csv'), 'mantenimiento_personal');
        
        // 11. Mantenimiento Insumos
        await importCSV(conn, join(dataDir, '11_mantenimiento_insumos.csv'), 'mantenimiento_insumos');
        
        // 12. Mantenimiento Actividades
        await importCSV(conn, join(dataDir, '12_mantenimiento_actividades.csv'), 'mantenimiento_actividades', ['actividad_id', 'observaciones']);
        
        await conn.execute('SET FOREIGN_KEY_CHECKS = 1');
        
        // Verificar datos
        console.log('');
        console.log('📊 Verificando datos cargados...');
        const [results] = await conn.execute(`
            SELECT 'Maquinaria' as tabla, COUNT(*) as registros FROM maquinaria
            UNION ALL SELECT 'Personal', COUNT(*) FROM personal
            UNION ALL SELECT 'Herramientas', COUNT(*) FROM herramientas
            UNION ALL SELECT 'Insumos', COUNT(*) FROM insumos
            UNION ALL SELECT 'Planes', COUNT(*) FROM planes_mantenimiento
            UNION ALL SELECT 'Actividades', COUNT(*) FROM actividades_mantenimiento
            UNION ALL SELECT 'Mantenimientos', COUNT(*) FROM mantenimientos
        `);
        
        results.forEach(row => {
            console.log(`   ${row.tabla}: ${row.registros} registros`);
        });
        
        await conn.end();
        
        console.log('');
        console.log('✅ Importación completada!');
    } catch (error) {
        console.error('❌ Error:', error.message);
        if (conn) await conn.end();
        process.exit(1);
    }
}

main();

